x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/ApproximationAndVerificationQuery.java

```diff
@@ -0,0 +1,250 @@
+package org.elasticsearch.xpack.wildcard.mapper;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.ConjunctionDISI;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Matches;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.ScorerSupplier;
+import org.apache.lucene.search.TwoPhaseIterator;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.search.Scorable.ChildScorable;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A wrapper for two query clauses, one a cheap approximation and the other a 
+ * more expensive verification. (The same principle as chainsaw and fine sandpaper,
+ * one quickly removes the bulk of what's needed but the other is required for 
+ * producing a final finish).
+ * Wrapping the clauses in a BooleanQuery as two "musts" rather than using this class 
+ * also works but there's an important disadvantage - the caching framework harvests 
+ * the internals of BooleanQueries, trying to find expensive clauses to cache. 
+ * Because verification clauses (the fine sandpaper) only ever get used in conjunction 
+ * with the results from a rough approximation we want to prevent them being used by 
+ * the query caching logic on the whole index (the equivalent of turning a large tree 
+ * into a food bowl only using fine sandpaper). 
+ * Equally, the approximation query provides inconclusive results that need further 
+ * treatment, so this clause would never be used in isolation (you don't make a food bowl
+ * using just a chainsaw).
+ * 
+ * This ApproximationAndVerificationQuery class provides
+ * an opaque wrapper whose results can be cached as a whole but whose component parts 
+ * cannot be visible to the caching framework. 
+ */
+public final class ApproximationAndVerificationQuery extends Query {
+
+    private final Query approxQuery, verifyQuery;
+
+    /**
+     * Create an {@link ApproximationAndVerificationQuery}. 
+     * @param approxQuery a query that is quick to run across many documents but is only approximate
+     * @param verifyQuery a query used to verify matches (slowly) on individual matches from approxQuery
+     */
+    public ApproximationAndVerificationQuery(Query approxQuery, Query verifyQuery) {
```

Comment:
I'm not sure adding this new class is super useful at the moment -- we only have one example right now that fits into this mold. I think it'd be best to just make a targeted fix for `AutomatonQueryOnBinaryDv` as a first iteration. Specifically, maybe this query could take in an approximation query as a parameter, and be renamed to something like `BinaryDvConfirmedAutomatonQuery`?

I suspect that a targeted fix will be clearer and have a lot less boilerplate. We can always follow-up later with a refactor if we encounter more examples of approximation + verification queries?